src/nvim/path.c

```diff
@@ -494,6 +494,275 @@ bool path_has_wildcard(const char_u *p)
+  // directory at a time from the tail of the path.
+  for (;;) {
+    // go back one path-separator
+    while (endp > save_fname && !after_pathsep(save_fname, endp + 1)) {
+      endp--;
+    }
+    if (endp <= save_fname) {
+      break;  // processed the complete path
+    }
+
+    // Replace the path separator with a NUL and try to shorten the
+    // resulting path.
+    char ch = *endp;
+    *endp = NUL;
+    if (os_path_exists((const char_u *)save_fname)) {
+      short_fname = path_to_short_save(save_fname);
+      if (short_fname == NULL) {
+        retval = false;
+        goto theend;
+      }
+      *endp = ch;  // preserve the string
+      retval = true;
+      break;  // successfully shortened the path
+    }
+    // failed to shorten the path. Skip the path separator
+    endp--;
+  }
+
+  size_t len = STRLEN(short_fname);
+  if (retval) {
+    // Succeeded in shortening the path. Now concatenate the shortened
+    // path with the remaining path at the tail.
+
+    // Compute the length of the new path.
+    size_t sfx_len = (size_t)(save_endp - endp + 1);
+    size_t new_len = len + sfx_len;
+
+    xfree(*bufp);
+    if (new_len > old_len) {
+      // There is not enough space in the currently allocated string,
+      // copy it to a buffer big enough.
+      *fname = *bufp = (char *)vim_strnsave((char_u *)short_fname, new_len);
+    } else {
+      // Transfer short_fname to the main buffer (it's big enough),
+      // unless get_short_pathname() did its work in-place.
+      *fname = *bufp = save_fname;
+      xstrlcpy(save_fname, short_fname, len + 1);
+      save_fname = NULL;
+    }
+
+    // concat the not-shortened part of the path
+    xstrlcpy(*fname + len, endp, sfx_len + 1);
+    (*fname)[new_len] = NUL;
+  }
+
+theend:
+  xfree(short_fname);
+  xfree(save_fname);
+
+  return retval;
+}
+
+// Get a pathname for a partial path.
+// Returns true for success, false for failure.
+bool path_short_for_partial(char **fnamep, char **bufp, size_t fnamelen)
+{
+  int sepcount;
+  char *p;
+
+  // Count up the path separators from the RHS.. so we know which part
+  // of the path to return.
+  sepcount = 0;
+  for (p = *fnamep; p < *fnamep + fnamelen; MB_PTR_ADV(p)) {
+    if (vim_ispathsep(*p)) {
+      sepcount++;
+    }
+  }
+
+  // Need full path first (use expand_env() to remove a "~/")
+  bool hasTilde = (**fnamep == '~');
+  char *pbuf, *tfname;
+  if (hasTilde) {
+    pbuf = tfname = (char *)expand_env_save((char_u *)(*fnamep));
+  } else {
+    pbuf = tfname = FullName_save(*fnamep, false);
+  }
+
+  size_t len = STRLEN(tfname);
+
+  if (os_path_exists((const char_u *)tfname)) {
+    if (!path_to_short(tfname, pbuf, len)) {
+      return false;
+    }
+  } else {
+    // Don't have a valid filename, so shorten the rest of the
+    // path if we can. This CAN give us invalid 8.3 filenames, but
+    // there's not a lot of point in guessing what it might be.
+    if (!path_short_for_invalid_fname(&tfname, &pbuf, len)) {
```

Comment:
This will never happen on a modern Windows system. So we can eliminate `path_short_for_invalid_fname` entirely.

If `os_path_exists` fails, the path doesn't exist. Shortening the filename won't help.

Reply:
vim returns the original path if:

```
:echo fnamemodif('C:\PROGRA~1\nvim\share\nvim\runtime\INVALI~1.VIM', ':p')
 => C:\PROGRA~1\nvim\share\nvim\runtime\INVALI~1.VIM'
```
However, in the following cases, the part that can be converted to the short file name is converted to the short file name.

```
:echo fnamemodif('C:\Program Files\nvim\share\nvim\runtime\invalid_path.vim', ':8')
 => C:\PROGRA~1\nvim\share\nvim\runtime\invalid_path.vim'
```

I don't know why this is implemented. However, I think `path_short_for_invalid_fname()` is necessary if the same behavior as vim is necessary.